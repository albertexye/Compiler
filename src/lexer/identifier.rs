use super::*;

impl Lexer {
    pub(super) fn read_identifier(&mut self) -> TokenValue {
        let mut identifier = Vec::new();
        while let Some(&ch) = self.peek()
            && (ch.is_alphanumeric() || ch == '_')
        {
            identifier.push(ch);
            self.advance();
        }
        let identifier = identifier.iter().collect();
        let id = self.symbol_table.insert(identifier);
        if SymbolTable::is_keyword(&id) {
            TokenValue::Keyword(SymbolTable::get_keyword(&id).unwrap())
        } else {
            TokenValue::Identifier(id)
        }
    }
}
