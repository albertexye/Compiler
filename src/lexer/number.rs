use super::*;

impl Lexer {
    /// Reads a number token (dispatches to decimal, float, hex, or binary).
    pub(crate) fn read_number(&mut self) -> Result<TokenValue, Error> {
        let ch = *self.peek().unwrap();
        if ch == '0'
            && let Some(&next_ch) = self.input.get(self.index + 1)
        {
            if next_ch == 'x' || next_ch == 'X' {
                return self.read_hexadecimal_number();
            } else if next_ch == 'b' || next_ch == 'B' {
                return self.read_binary_number();
            }
        }
        self.read_decimal_or_float_number()
    }

    fn read_decimal_or_float_number(&mut self) -> Result<TokenValue, Error> {
        let negative = self.consume_negative_sign();
        let number = self.collect_digits()?;
        if self.peek() != Some(&'.') {
            return self.make_integer(number, negative);
        }
        self.advance(); // skip '.'
        let fraction = self.collect_fraction()?;
        let float_value = (number as f64) + fraction;
        let final_value = if negative { -float_value } else { float_value };
        Ok(TokenValue::Literal(Literal::Float(final_value)))
    }

    fn consume_negative_sign(&mut self) -> bool {
        if *self.peek().unwrap() == '-' {
            self.advance();
            true
        } else {
            false
        }
    }

    fn collect_digits(&mut self) -> Result<u64, Error> {
        let mut number = 0u64;
        let mut found = false;
        while let Some(&ch) = self.peek() {
            if !ch.is_ascii_digit() {
                break;
            }
            found = true;
            number = number
                .checked_mul(10)
                .and_then(|n| n.checked_add(ch.to_digit(10).unwrap() as u64))
                .ok_or_else(|| {
                    self.error(
                        ErrorType::InvalidNumber,
                        "Integer overflow in number".to_string(),
                    )
                })?;
            self.advance();
        }
        if !found {
            return Err(self.error(
                ErrorType::InvalidNumber,
                "No digits found in number".to_string(),
            ));
        }
        Ok(number)
    }

    fn collect_fraction(&mut self) -> Result<f64, Error> {
        let mut fraction = 0f64;
        let mut divisor = 10f64;
        let mut found = false;
        while let Some(&ch) = self.peek() {
            if !ch.is_ascii_digit() {
                break;
            }
            found = true;
            fraction += (ch.to_digit(10).unwrap() as f64) / divisor;
            divisor *= 10f64;
            self.advance();
        }
        if !found {
            return Err(self.error(
                ErrorType::InvalidNumber,
                "No digits found after decimal point".to_string(),
            ));
        }
        Ok(fraction)
    }

    fn make_integer(&self, number: u64, negative: bool) -> Result<TokenValue, Error> {
        if negative {
            if number - 1 > i64::MAX as u64 {
                Err(self.error(
                    ErrorType::InvalidNumber,
                    "Integer overflow in negative number".to_string(),
                ))
            } else {
                Ok(TokenValue::Literal(Literal::Int(-(number as i64))))
            }
        } else {
            Ok(TokenValue::Literal(Literal::UInt(number)))
        }
    }

    fn read_hexadecimal_number(&mut self) -> Result<TokenValue, Error> {
        self.advance(); // skip '0'
        self.advance(); // skip 'x' or 'X'
        let mut hex_str = String::new();
        while let Some(&ch) = self.peek() {
            if ch.is_ascii_hexdigit() {
                hex_str.push(ch);
                self.advance();
            } else {
                break;
            }
        }
        if hex_str.is_empty() {
            return Err(self.error(
                ErrorType::InvalidNumber,
                "No digits found in hexadecimal number".to_string(),
            ));
        }
        if let Ok(value) = u64::from_str_radix(&hex_str, 16) {
            Ok(TokenValue::Literal(Literal::UInt(value)))
        } else {
            Err(self.error(
                ErrorType::InvalidNumber,
                format!("Invalid hexadecimal number: 0x{}", hex_str),
            ))
        }
    }

    fn read_binary_number(&mut self) -> Result<TokenValue, Error> {
        self.advance(); // skip '0'
        self.advance(); // skip 'b' or 'B'
        let mut bin_str = String::new();
        while let Some(&ch) = self.peek() {
            if ch == '0' || ch == '1' {
                bin_str.push(ch);
                self.advance();
            } else {
                break;
            }
        }
        if bin_str.is_empty() {
            return Err(self.error(
                ErrorType::InvalidNumber,
                "No digits found in binary number".to_string(),
            ));
        }
        if let Ok(value) = u64::from_str_radix(&bin_str, 2) {
            Ok(TokenValue::Literal(Literal::UInt(value)))
        } else {
            Err(self.error(
                ErrorType::InvalidNumber,
                format!("Invalid binary number: 0b{}", bin_str),
            ))
        }
    }
}
