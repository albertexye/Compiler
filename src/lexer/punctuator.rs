use super::*;

impl Lexer {
    pub(super) fn read_punctuator(&mut self) -> Result<TokenValue, Error> {
        let mut punc = Vec::new();
        let mut keyword = None;
        let mut kw_i = 0usize;
        let mut i = 0usize;
        while let Some(ch) = self.input.get(self.index + i)
            && ch.is_ascii_punctuation()
        {
            punc.push(*ch);
            let s: String = punc.iter().collect();
            if let Some(id) = self.symbol_table.search(&s) {
                if SymbolTable::is_keyword(&id) {
                    keyword = SymbolTable::get_keyword(&id);
                    kw_i = i;
                }
                i += 1;
            } else {
                break;
            }
        }
        if let Some(kw) = keyword {
            self.index += kw_i + 1;
            self.column += kw_i + 1;
            Ok(TokenValue::Keyword(kw))
        } else {
            Err(self.error(ErrorType::UnknownCharacter, "Unknown punctuator"))
        }
    }
}
