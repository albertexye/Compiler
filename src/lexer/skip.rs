use super::*;

impl Lexer {
    pub(crate) fn skip_whitespace_and_comments(&mut self) {
        while self.skip_whitespace() || self.skip_comment() {}
    }

    fn skip_whitespace(&mut self) -> bool {
        let mut found = false;
        while let Some(&ch) = self.peek() {
            if !ch.is_whitespace() {
                break;
            }
            found = true;
            self.advance();
        }
        found
    }

    fn skip_comment(&mut self) -> bool {
        if self.peek() == Some(&'/') && self.peek2() == Some(&'/') {
            while let Some(&ch) = self.peek() {
                self.advance();
                if ch == '\n' {
                    break;
                }
            }
            true
        } else {
            false
        }
    }
}
