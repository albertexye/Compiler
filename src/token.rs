use serde::Serialize;
use std::collections::HashMap;
use std::ops::Sub;
#[cfg(test)]
use std::sync::{LazyLock, Mutex};

#[derive(Debug, PartialEq, Eq, Clone, Copy)]
pub(crate) enum TokenType {
    Comma,
    Semicolon,
    Colon,
    DoubleColon,
    Dot,
    OpenParen,
    CloseParen,
    OpenBrace,
    CloseBrace,
    OpenBracket,
    CloseBracket,
    Plus,
    PlusEq,
    Minus,
    MinusEq,
    Mul,
    MulEq,
    Div,
    DivEq,
    Modulo,
    ModuloEq,
    LeftShift,
    LeftShiftEq,
    RightShift,
    RightShiftEq,
    BitAnd,
    BitAndEq,
    BitOr,
    BitOrEq,
    BitXor,
    BitXorEq,
    BitNot,
    LogicalAnd,
    LogicalOr,
    LogicalNot,
    Eq,
    NotEq,
    Gt,
    Ge,
    Lt,
    Le,
    Assign,
    ReturnType,
    MatchCase,
    If,
    Else,
    Match,
    While,
    For,
    Break,
    Continue,
    Return,
    Fn,
    Let,
    Var,
    Struct,
    Enum,
    Union,
    Pub,
    Prv,
    Mod,
    Module,
    Import,
    Use,
}

#[derive(Debug, PartialEq, Clone)]
pub(crate) enum Literal {
    UInt(u64),
    Int(i64),
    Float(f64),
    String(String),
}

#[derive(Debug, PartialEq, Clone)]
pub(crate) enum TokenValue {
    Identifier(SymbolId),
    Literal(Literal),
    Keyword(TokenType),
}

#[derive(Debug, PartialEq, Clone, Copy, Serialize)]
pub(crate) struct TokenSpan {
    pub(crate) line: usize,
    pub(crate) column: usize,
    pub(crate) index: usize,
    pub(crate) size: usize,
}

#[derive(Debug, PartialEq, Clone)]
pub(crate) struct Token {
    pub(crate) value: TokenValue,
    pub(crate) span: TokenSpan,
}

impl Sub for TokenSpan {
    type Output = TokenSpan;

    fn sub(self, other: TokenSpan) -> TokenSpan {
        std::debug_assert!(self.index + self.size >= other.index);
        TokenSpan {
            line: self.line,
            column: self.column,
            index: self.index,
            size: self.index + self.size - other.index,
        }
    }
}

const TOKEN_TYPES_STR: [&str; 64] = [
    ",", ";", ":", "::", ".", "(", ")", "[", "]", "{", "}", "+", "+=", "-", "-=", "*", "*=", "/",
    "/=", "%", "%=", "<<", "<<=", ">>", ">>=", "&", "&=", "|", "|=", "^", "^=", "~", "and", "or",
    "!", "==", "!=", ">", ">=", "<", "<=", "=", "->", "=>", "if", "else", "match", "while", "for",
    "break", "continue", "return", "fn", "let", "var", "struct", "enum", "union", "pub", "prv",
    "mod", "module", "import", "use",
];

const TOKEN_TYPES_ENUM: [TokenType; 64] = [
    TokenType::Comma,
    TokenType::Semicolon,
    TokenType::Colon,
    TokenType::DoubleColon,
    TokenType::Dot,
    TokenType::OpenParen,
    TokenType::CloseParen,
    TokenType::OpenBrace,
    TokenType::CloseBrace,
    TokenType::OpenBracket,
    TokenType::CloseBracket,
    TokenType::Plus,
    TokenType::PlusEq,
    TokenType::Minus,
    TokenType::MinusEq,
    TokenType::Mul,
    TokenType::MulEq,
    TokenType::Div,
    TokenType::DivEq,
    TokenType::Modulo,
    TokenType::ModuloEq,
    TokenType::LeftShift,
    TokenType::LeftShiftEq,
    TokenType::RightShift,
    TokenType::RightShiftEq,
    TokenType::BitAnd,
    TokenType::BitAndEq,
    TokenType::BitOr,
    TokenType::BitOrEq,
    TokenType::BitXor,
    TokenType::BitXorEq,
    TokenType::BitNot,
    TokenType::LogicalAnd,
    TokenType::LogicalOr,
    TokenType::LogicalNot,
    TokenType::Eq,
    TokenType::NotEq,
    TokenType::Gt,
    TokenType::Ge,
    TokenType::Lt,
    TokenType::Le,
    TokenType::Assign,
    TokenType::ReturnType,
    TokenType::MatchCase,
    TokenType::If,
    TokenType::Else,
    TokenType::Match,
    TokenType::While,
    TokenType::For,
    TokenType::Break,
    TokenType::Continue,
    TokenType::Return,
    TokenType::Fn,
    TokenType::Let,
    TokenType::Var,
    TokenType::Struct,
    TokenType::Enum,
    TokenType::Union,
    TokenType::Pub,
    TokenType::Prv,
    TokenType::Mod,
    TokenType::Module,
    TokenType::Import,
    TokenType::Use,
];

const _: () = assert!(TOKEN_TYPES_STR.len() == TOKEN_TYPES_ENUM.len());

#[cfg_attr(not(test), derive(Serialize))]
#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash)]
pub(crate) struct SymbolId(usize);

pub(crate) struct SymbolTable {
    counter: SymbolId,
    table: HashMap<String, SymbolId>,
    reverse: Option<Vec<String>>,
}

impl SymbolTable {
    pub(crate) fn new() -> SymbolTable {
        let mut sym_table = SymbolTable {
            counter: SymbolId(0),
            table: HashMap::new(),
            reverse: None,
        };
        for keyword in TOKEN_TYPES_STR {
            sym_table
                .table
                .insert(keyword.to_string(), sym_table.counter);
            sym_table.counter.0 += 1;
        }
        sym_table
    }

    pub(crate) fn insert(&mut self, token: String) -> SymbolId {
        std::debug_assert!(self.reverse.is_none());
        if self.table.contains_key(&token) {
            self.table[&token]
        } else {
            let id = self.counter;
            self.table.insert(token, self.counter);
            self.counter.0 += 1;
            id
        }
    }

    pub(crate) fn search(&self, token: &str) -> Option<SymbolId> {
        std::debug_assert!(self.reverse.is_none());
        if self.table.contains_key(token) {
            Some(self.table[token])
        } else {
            None
        }
    }

    pub(crate) fn is_keyword(id: &SymbolId) -> bool {
        id.0 < TOKEN_TYPES_STR.len()
    }

    pub(crate) fn get_keyword(id: &SymbolId) -> Option<TokenType> {
        if !SymbolTable::is_keyword(id) {
            None
        } else {
            Some(TOKEN_TYPES_ENUM[id.0])
        }
    }

    pub(crate) fn reverse_lookup(&mut self, id: SymbolId) -> String {
        let rev = match self.reverse.as_ref() {
            Some(rev) => rev,
            None => {
                let mut reverse = vec![String::new(); self.counter.0];
                let table = std::mem::take(&mut self.table);
                for (sym, id) in table.into_iter() {
                    reverse[id.0] = sym;
                }
                self.reverse = Some(reverse);
                self.reverse.as_ref().unwrap()
            }
        };
        assert!(id.0 < rev.len());
        rev[id.0].clone()
    }
}

#[cfg(test)]
static SYMBOL_CONTEXT: LazyLock<Mutex<SymbolTable>> =
    LazyLock::new(|| Mutex::new(SymbolTable::new()));

#[cfg(test)]
pub(crate) fn set_symbol_context(symbol_table: SymbolTable) {
    let mut guard = SYMBOL_CONTEXT.lock().unwrap();
    *guard = symbol_table;
}

#[cfg(test)]
impl Serialize for SymbolId {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: serde::Serializer,
    {
        let mut guard = SYMBOL_CONTEXT.lock().unwrap();
        serializer.serialize_str(&guard.reverse_lookup(*self))
    }
}
