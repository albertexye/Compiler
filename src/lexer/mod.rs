use crate::token::*;

mod identifier;
mod number;
mod punctuator;
mod skip;
mod string;
mod utils;

pub(crate) struct Lexer {
    input: Vec<char>,

    index: usize,
    line: usize,
    column: usize,

    start_index: usize,
    start_line: usize,
    start_column: usize,

    pub(crate) symbol_table: SymbolTable,
}

#[derive(Debug, PartialEq, Eq, Clone, Copy)]
pub(crate) enum ErrorType {
    UnclosedString,
    InvalidEscapeSequence,
    InvalidNumber,
    UnknownCharacter,
}

#[derive(Debug, PartialEq, Clone)]
pub(crate) struct Error {
    typ: ErrorType,
    span: TokenSpan,
    msg: &'static str,
}

impl Lexer {
    pub(crate) fn new() -> Self {
        Lexer {
            input: Vec::new(),
            index: 0,
            line: 1,
            column: 1,
            start_index: 0,
            start_line: 1,
            start_column: 1,
            symbol_table: SymbolTable::new(),
        }
    }

    pub(crate) fn lex(&mut self, input: &str) -> Result<Vec<Token>, Error> {
        self.input = input.chars().collect();
        self.index = 0;
        self.line = 1;
        self.column = 1;
        self.start_index = 0;
        self.start_line = 1;
        self.start_column = 1;
        let mut tokens = Vec::new();
        while let Some(token) = self.next_token()? {
            tokens.push(token);
        }
        self.input = Vec::new();
        Ok(tokens)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::token::{Literal, Token, TokenSpan, TokenValue};

    // Helper to create a TokenSpan
    fn span(line: usize, column: usize, index: usize, size: usize) -> TokenSpan {
        TokenSpan {
            line,
            column,
            index,
            size,
        }
    }

    // Helper function to compare lexed tokens with expected tokens, including spans.
    fn assert_lexes(input: &str, expected: Vec<Token>) {
        let mut lexer = Lexer::new();
        let tokens = lexer.lex(input).unwrap();
        assert_eq!(tokens, expected);
    }

    #[test]
    fn test_empty_input() {
        assert_lexes("", vec![]);
    }

    #[test]
    fn test_whitespace_and_comments() {
        assert_lexes("  // this is a comment\n  ", vec![]);
        assert_lexes("// another comment", vec![]);
    }

    #[test]
    fn test_integers() {
        assert_lexes(
            "123",
            vec![Token {
                value: TokenValue::Literal(Literal::UInt(123)),
                span: span(1, 1, 0, 3),
            }],
        );
        assert_lexes(
            "-45",
            vec![Token {
                value: TokenValue::Literal(Literal::Int(-45)),
                span: span(1, 1, 0, 3),
            }],
        );
    }

    #[test]
    fn test_hexadecimal_numbers() {
        assert_lexes(
            "0x1A",
            vec![Token {
                value: TokenValue::Literal(Literal::UInt(26)),
                span: span(1, 1, 0, 4),
            }],
        );
        assert_lexes(
            "0Xff",
            vec![Token {
                value: TokenValue::Literal(Literal::UInt(255)),
                span: span(1, 1, 0, 4),
            }],
        );
    }

    #[test]
    fn test_binary_numbers() {
        assert_lexes(
            "0b1010",
            vec![Token {
                value: TokenValue::Literal(Literal::UInt(10)),
                span: span(1, 1, 0, 6),
            }],
        );
    }

    #[test]
    fn test_float_numbers() {
        assert_lexes(
            "123.456",
            vec![Token {
                value: TokenValue::Literal(Literal::Float(123.456)),
                span: span(1, 1, 0, 7),
            }],
        );
        assert_lexes(
            "-0.5",
            vec![Token {
                value: TokenValue::Literal(Literal::Float(-0.5)),
                span: span(1, 1, 0, 4),
            }],
        );
    }

    #[test]
    fn test_number_errors() {
        let mut lexer = Lexer::new();
        assert!(lexer.lex("0xG").is_err());
        assert!(lexer.lex("0b2").is_err());
    }

    #[test]
    fn test_string_literals() {
        assert_lexes(
            r#""hello""#,
            vec![Token {
                value: TokenValue::Literal(Literal::String("hello".to_string())),
                span: span(1, 1, 0, 7),
            }],
        );
        assert_lexes(
            r#""escaped \" \n \t \\""#,
            vec![Token {
                value: TokenValue::Literal(Literal::String("escaped \" \n \t \\".to_string())),
                span: span(1, 1, 0, 21),
            }],
        );
    }

    #[test]
    fn test_unclosed_string() {
        let mut lexer = Lexer::new();
        assert!(lexer.lex(r#""hello"#).is_err());
    }

    #[test]
    fn test_identifiers_and_keywords() {
        let mut lexer = Lexer::new();
        let tokens = lexer.lex("let x = 5;").unwrap();
        let expected = vec![
            Token {
                value: TokenValue::Keyword(TokenType::Let),
                span: span(1, 1, 0, 3),
            },
            Token {
                value: TokenValue::Identifier(lexer.symbol_table.search("x").unwrap()),
                span: span(1, 5, 4, 1),
            },
            Token {
                value: TokenValue::Keyword(TokenType::Assign),
                span: span(1, 7, 6, 1),
            },
            Token {
                value: TokenValue::Literal(Literal::UInt(5)),
                span: span(1, 9, 8, 1),
            },
            Token {
                value: TokenValue::Keyword(TokenType::Semicolon),
                span: span(1, 10, 9, 1),
            },
        ];
        assert_eq!(tokens, expected);
    }

    #[test]
    fn test_punctuators() {
        assert_lexes(
            "+ - * /",
            vec![
                Token {
                    value: TokenValue::Keyword(TokenType::Plus),
                    span: span(1, 1, 0, 1),
                },
                Token {
                    value: TokenValue::Keyword(TokenType::Minus),
                    span: span(1, 3, 2, 1),
                },
                Token {
                    value: TokenValue::Keyword(TokenType::Mul),
                    span: span(1, 5, 4, 1),
                },
                Token {
                    value: TokenValue::Keyword(TokenType::Div),
                    span: span(1, 7, 6, 1),
                },
            ],
        );
    }

    #[test]
    fn test_multiline_lexing() {
        let mut lexer = Lexer::new();
        let tokens = lexer.lex("let y\n  = 10;").unwrap();
        let expected = vec![
            Token {
                value: TokenValue::Keyword(TokenType::Let),
                span: span(1, 1, 0, 3),
            },
            Token {
                value: TokenValue::Identifier(lexer.symbol_table.search("y").unwrap()),
                span: span(1, 5, 4, 1),
            },
            Token {
                value: TokenValue::Keyword(TokenType::Assign),
                span: span(2, 3, 8, 1),
            },
            Token {
                value: TokenValue::Literal(Literal::UInt(10)),
                span: span(2, 5, 10, 2),
            },
            Token {
                value: TokenValue::Keyword(TokenType::Semicolon),
                span: span(2, 7, 12, 1),
            },
        ];
        assert_eq!(tokens, expected);
    }
}
