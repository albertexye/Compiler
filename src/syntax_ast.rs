use std::{
    collections::{HashMap, HashSet},
    path::PathBuf,
};

use crate::token::TokenSpan;

pub(crate) struct Ast {
    pub(crate) entry: PathBuf,
    pub(crate) modules: HashMap<PathBuf, Module>,
}

pub(crate) struct Module {
    pub(crate) name: String,
    pub(crate) files: HashMap<String, File>,
}

pub(crate) struct File {
    pub(crate) name: String,
    pub(crate) module: String,
    pub(crate) imports: HashSet<String>,
    pub(crate) globals: HashMap<String, Scope<Declaration>>,
    pub(crate) functions: HashMap<String, Scope<Function>>,
    pub(crate) types: HashMap<String, Scope<TypeDef>>,
}

pub(crate) type Name = Vec<String>;

pub(crate) enum Visibility {
    Public,
    Private,
    Module,
}

pub(crate) struct Scope<T> {
    pub(crate) visibility: Visibility,
    pub(crate) value: T,
}

pub(crate) struct TypeDef {
    pub(crate) name: String,
    pub(crate) body: TypeDefBody,
    pub(crate) span: TokenSpan,
}

pub(crate) enum TypeDefBody {
    Struct(HashMap<String, TypeAnnot>),
    Enum(HashMap<String, u64>),
    Union(HashMap<String, TypeAnnot>),
    Alias(TypeAnnot),
}

pub(crate) enum Statement {
    Declaration(Declaration),
    Assignment(Assignment),
    Expression(Expression),
    Loop(Loop),
    Continue(TokenSpan),
    Break(TokenSpan),
    Conditional(Conditional),
    Match(Match),
    Return(Expression),
}

pub(crate) struct TypeAnnot {
    pub(crate) base: Name,
    pub(crate) modifiers: Vec<TypeModifier>,
    pub(crate) span: TokenSpan,
}

pub(crate) struct TypeModifier {
    pub(crate) mutable: bool,
    pub(crate) typ: TypeModifierType,
}

pub(crate) enum TypeModifierType {
    Pointer,
    Slice,
    Array(u64),
}

pub(crate) struct Function {
    pub(crate) name: String,
    pub(crate) arguments: Vec<FunctionArg>,
    pub(crate) return_type: Option<TypeAnnot>,
    pub(crate) body: Vec<Statement>,
    pub(crate) span: TokenSpan,
}

pub(crate) struct FunctionArg {
    pub(crate) name: String,
    pub(crate) typ: TypeAnnot,
    pub(crate) span: TokenSpan,
}

pub(crate) struct Declaration {
    pub(crate) name: String,
    pub(crate) mutable: bool,
    pub(crate) typ: TypeAnnot,
    pub(crate) value: Expression,
    pub(crate) span: TokenSpan,
}

pub(crate) struct Expression {
    pub(crate) value: ExpressionValue,
    pub(crate) span: TokenSpan,
}

pub(crate) enum ExpressionValue {
    Binary(Binary),
    Unary(Unary),
    Call(Call),
    Literal(Literal),
    Identifier(Name),
}

pub(crate) struct Binary {
    pub(crate) left: Box<Expression>,
    pub(crate) right: Box<Expression>,
    pub(crate) op: BinaryOp,
}

pub(crate) enum BinaryOp {
    Plus,
    Minus,
    Mul,
    Div,
    Mod,
    LeftShift,
    RightShift,
    BitAnd,
    BitOr,
    BitXor,
    Gt,
    Ge,
    Lt,
    Le,
    Eq,
    NotEq,
    LogicalAnd,
    LogicalOr,
    Indexing,
    FieldAccess,
}

pub(crate) struct Unary {
    pub(crate) operand: Box<Expression>,
    pub(crate) op: UnaryOp,
}

pub(crate) enum UnaryOp {
    LogicalNot,
    BitNot,
    Dereference,
    AddressOf,
    Negate,
}

pub(crate) struct Call {
    pub(crate) function: Box<Expression>,
    pub(crate) args: Vec<Expression>,
}

pub(crate) enum Literal {
    String(String),
    UInt(u64),
    Int(i64),
    Float(f64),
    Bool(bool),
    Array(Vec<Expression>),
    Struct(HashMap<String, Expression>),
}

pub(crate) struct Assignment {
    pub(crate) left: Expression,
    pub(crate) right: Expression,
    pub(crate) typ: AssignmentType,
    pub(crate) span: TokenSpan,
}

pub(crate) enum AssignmentType {
    Assign,
    Plus,
    Minus,
    Mul,
    Div,
    Mod,
    LeftShift,
    RightShift,
    BitAnd,
    BitOr,
    BitXor,
}

pub(crate) struct ConditionalBranch {
    pub(crate) condition: Expression,
    pub(crate) body: Vec<Statement>,
}

pub(crate) struct Conditional {
    pub(crate) if_branch: ConditionalBranch,
    pub(crate) elif_branches: Vec<ConditionalBranch>,
    pub(crate) else_branch: Option<Vec<Statement>>,
}

pub(crate) struct Match {
    pub(crate) value: Expression,
    pub(crate) cases: Vec<ConditionalBranch>,
    pub(crate) default: Option<Vec<Statement>>,
}

pub(crate) struct Loop {
    pub(crate) init: Option<Declaration>,
    pub(crate) condition: Option<Expression>,
    pub(crate) update: Vec<Statement>,
    pub(crate) body: Vec<Statement>,
}
