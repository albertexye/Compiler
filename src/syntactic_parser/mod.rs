use crate::syntax_ast;
use crate::syntax_ast::{Name, Statement};
use crate::token;
use crate::token::{Token, TokenType, TokenValue};

mod assignment;
mod conditional;
mod declaration;
mod expression;
mod file;
mod function;
mod r#loop;
mod r#match;
mod r#return;
mod statement;
mod type_annotation;
mod type_definition;
mod utils;

#[derive(Debug)]
pub(crate) enum ErrorType {
    Name,
    Module,
    Import,
    LineEnd,
    TypeDefinition,
    Declaration,
    TypeAnnotation,
    Expression,
    Statement,
    Conditional,
    Function,
    Match,
    Loop,
}

#[derive(Debug)]
pub(crate) struct Error {
    typ: ErrorType,
    msg: String,
    token: Option<Token>,
}

pub struct SyntacticParser {
    tokens: Vec<Token>,
    index: usize,
}

impl SyntacticParser {
    pub(crate) fn parse(
        tokens: Vec<Token>,
        filename: &str,
        module_name: &str,
    ) -> Result<syntax_ast::File, Error> {
        let mut parser = SyntacticParser {
            tokens,
            index: 0usize,
        };
        parser.parse_file(filename, module_name)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::lexer::Lexer;
    use crate::syntax_ast::File;
    use insta;

    fn parse_test_file(code: &str, filename: &str, module_name: &str) -> File {
        let tokens = Lexer::lex(code).unwrap();
        SyntacticParser::parse(tokens, filename, module_name).unwrap()
    }

    #[test]
    fn test_basic() {
        let code = r#"module test_add;

import std;

prv fn add(a: i32, b: i32) -> i32 {
    let ret: i32 = a + b;
    return ret;
}

pub fn test() -> bool {
    let expected: i32 = 25;
    let result: i32 = add(30, -5);
    if (result == expected) {
        std::print("Passed!\n");
        return true;
    } else {
        std::print("Failed!\n");
        return false;
    }
}"#;
        let ast = parse_test_file(code, "test", "test_add");

        let mut settings = insta::Settings::clone_current();
        settings.set_sort_maps(true);
        settings.bind(|| {
            insta::assert_yaml_snapshot!(ast);
        });
    }
}
