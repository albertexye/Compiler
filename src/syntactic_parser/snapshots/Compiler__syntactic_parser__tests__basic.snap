---
source: src/syntactic_parser/mod.rs
expression: ast
---
name: test
module: test_add
imports:
  - std
globals: {}
functions:
  add:
    visibility: Private
    value:
      name: add
      arguments:
        - name: a
          typ:
            base:
              - i32
            modifiers: []
            span:
              line: 5
              column: 15
              index: 45
              size: 3
          span:
            line: 5
            column: 15
            index: 45
            size: 6
        - name: b
          typ:
            base:
              - i32
            modifiers: []
            span:
              line: 5
              column: 23
              index: 53
              size: 3
          span:
            line: 5
            column: 23
            index: 53
            size: 6
      return_type:
        base:
          - i32
        modifiers: []
        span:
          line: 5
          column: 31
          index: 61
          size: 3
      body:
        - Declaration:
            name: ret
            mutable: false
            typ:
              base:
                - i32
              modifiers: []
              span:
                line: 6
                column: 14
                index: 80
                size: 3
            value:
              value:
                Binary:
                  left:
                    value:
                      Identifier:
                        - a
                    span:
                      line: 6
                      column: 20
                      index: 86
                      size: 1
                  right:
                    value:
                      Identifier:
                        - b
                    span:
                      line: 6
                      column: 24
                      index: 90
                      size: 1
                  op: Plus
              span:
                line: 6
                column: 24
                index: 90
                size: 3
            span:
              line: 6
              column: 25
              index: 91
              size: 21
        - Return:
            value:
              Identifier:
                - ret
            span:
              line: 7
              column: 12
              index: 104
              size: 3
      span:
        line: 5
        column: 8
        index: 38
        size: 3
  test:
    visibility: Public
    value:
      name: test
      arguments: []
      return_type:
        base:
          - bool
        modifiers: []
        span:
          line: 10
          column: 18
          index: 129
          size: 4
      body:
        - Declaration:
            name: expected
            mutable: false
            typ:
              base:
                - i32
              modifiers: []
              span:
                line: 11
                column: 19
                index: 154
                size: 3
            value:
              value:
                Literal:
                  UInt: 25
              span:
                line: 11
                column: 25
                index: 160
                size: 2
            span:
              line: 11
              column: 27
              index: 162
              size: 23
        - Declaration:
            name: result
            mutable: false
            typ:
              base:
                - i32
              modifiers: []
              span:
                line: 12
                column: 17
                index: 180
                size: 3
            value:
              value:
                Call:
                  function:
                    value:
                      Identifier:
                        - add
                    span:
                      line: 12
                      column: 23
                      index: 186
                      size: 3
                  args:
                    - value:
                        Literal:
                          UInt: 30
                      span:
                        line: 12
                        column: 27
                        index: 190
                        size: 2
                    - value:
                        Literal:
                          Int: -5
                      span:
                        line: 12
                        column: 31
                        index: 194
                        size: 2
              span:
                line: 12
                column: 33
                index: 196
                size: 8
            span:
              line: 12
              column: 34
              index: 197
              size: 30
        - Conditional:
            if_branch:
              condition:
                value:
                  Binary:
                    left:
                      value:
                        Identifier:
                          - result
                      span:
                        line: 13
                        column: 9
                        index: 207
                        size: 6
                    right:
                      value:
                        Identifier:
                          - expected
                      span:
                        line: 13
                        column: 19
                        index: 217
                        size: 8
                    op: Eq
                span:
                  line: 13
                  column: 19
                  index: 217
                  size: 11
              body:
                - Expression:
                    value:
                      Call:
                        function:
                          value:
                            Identifier:
                              - std
                              - print
                          span:
                            line: 14
                            column: 14
                            index: 242
                            size: 10
                        args:
                          - value:
                              Literal:
                                String: "Passed!\n"
                            span:
                              line: 14
                              column: 20
                              index: 248
                              size: 11
                    span:
                      line: 14
                      column: 31
                      index: 259
                      size: 13
                - Return:
                    value:
                      Identifier:
                        - "true"
                    span:
                      line: 15
                      column: 16
                      index: 277
                      size: 4
            elif_branches: []
            else_branch:
              - Expression:
                  value:
                    Call:
                      function:
                        value:
                          Identifier:
                            - std
                            - print
                        span:
                          line: 17
                          column: 14
                          index: 309
                          size: 10
                      args:
                        - value:
                            Literal:
                              String: "Failed!\n"
                          span:
                            line: 17
                            column: 20
                            index: 315
                            size: 11
                  span:
                    line: 17
                    column: 31
                    index: 326
                    size: 13
              - Return:
                  value:
                    Identifier:
                      - "false"
                  span:
                    line: 18
                    column: 16
                    index: 344
                    size: 5
      span:
        line: 10
        column: 8
        index: 119
        size: 4
types: {}
